---
layout: lesson
root: .  # Is the only page that doesn't follow the pattern /:path/index.html
permalink: index.html  # Is the only page that doesn't follow the pattern /:path/index.html
---

With the age of high-throughput sequencing, several techniques have been developed to genotype thousands of markers across hundreds of individuals at relatively low-cost. One group of those methods, Genotyping-by-Sequencing (GBS) cuts down on the amount of the genome sequenced by using restriction enzymes to repeatedly sequence the same genomic locations across individuals. The thousands of loci sequenced can bring high resolution to population genomic or phylogenomic analyses at a fraction of the cost associated with whole genome sequencing.

This one day workshop is designed to show how to process GBS/RAD data through hands-on analysis using the Stacks software pipeline.



<!-- this is an html comment -->

{% comment %} This is a comment in Liquid {% endcomment %}

> ## Prerequisites
>
> All pre-requisites are taken care of if you are using this lesson at the Otago Bioinformatics Spring School 2021.
>
> Please see the [setup](setup.md) instructions if you wish to undertake this workshop on your own computer.
{: .prereq}

{% include links.md %}
