---
title: Setup
---


This workshop is designed to be conducted on the [NeSI](https://www.nesi.org.nz) compute infrastructure. All software and data is already set up for you to use during the workshop.

Running the workshop locally on your own computer will involve installing the required programs and downloading the example data. Some software packages may not work on all operating systems. Below are the requirements for running locally.


## Software

The software needed to run this lesson are listed below. Versions reported are the ones we use on NeSI but the lesson is likely to work with a range of versions of these softwares.

| Software	| Version | 	Manual	| Description |
----------------------------------------------
| Stacks | 2.58| [link](https://catchenlab.life.illinois.edu/stacks) | GBS/RAD-Seq SNP calling pipeline. |
| BWA	0.7.17 |	[link](http://bio-bwa.sourceforge.net/bwa.shtml) |		Mapping DNA sequences. |
| SAMtools |	1.9	| [link](http://www.htslib.org/doc/samtools.html)		|Utilities for manipulating alignments in the SAM format. |
| R | 4.1.0 | [link](https://www.r-project.org/) | Statistical computing language. |
| Software	Version	Manual	Available for	Description |

The following R packages are alse required
R packages:
pcadapt
vcfR
pophelper


## Data

The files mentioned can be downloaded [here](https://drive.google.com/file/d/1x10Htq1Ddooh0AznHUtxlgPhZEDEFrys/view?usp=sharing).

The data used in this workshop comes from:

Ingram, T., Dutoit, L., Mikheev, P., Khan, S., & Schallenberg, M. (2020). Phenotypic, ecological, and genomic variation in common bully (Gobiomorphus cotidianus) populations along depth gradients in New Zealandâ€™s southern Great Lakes. Canadian Journal of Fisheries and Aquatic Sciences, 77(10), 1678-1687. [doi:0.1139/cjfas-2020-0015]( https://doi.org/10.1139/cjfas-2020-0015)


The files mentioned can be downloaded [here](https://drive.google.com/file/d/1x10Htq1Ddooh0AznHUtxlgPhZEDEFrys/view?usp=sharing).


{% include links.md %}
